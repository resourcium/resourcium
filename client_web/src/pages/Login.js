import React, { useCallback, useEffect, useState } from 'react';
import './Login.scss';
import logo from 'src/images/ucl-logo.jpg';
import * as Msal from 'msal';
import { useDispatch } from 'react-redux';
import { login } from '../state/actions';
import * as graph from '@microsoft/microsoft-graph-client';

import { AZURE_DETAILS } from 'src/common/config';

function getAccessToken(msalInstance) {
    const request = { scopes: AZURE_DETAILS.scopes };

    msalInstance.loginRedirect(request);
}

export async function getUserDetails(accessToken) {
    const client = graph.Client.init({
        authProvider: done => done(null, accessToken),
    });

    return await client.api('/me').get();
}

export default function Login() {
    const [msalInstance, setMsalInstance] = useState();
    const dispatch = useDispatch();

    const handleAccessToken = accessToken => {
        getUserDetails(accessToken)
            .then(user => {
                dispatch(login(user, accessToken));
                if (
                    !process.env.NODE_ENV ||
                    process.env.NODE_ENV === 'development'
                ) {
                    console.log('Access Token:', accessToken);
                }
            })
            .catch(err => {
                console.log('Error getting user', err);
            });
    };

    useEffect(() => {
        const msal = new Msal.UserAgentApplication({
            auth: {
                clientId: AZURE_DETAILS.appId,
                redirectUri: window.location.origin,
                postLogoutRedirectUri: window.location,
            },
            cache: {
                cacheLocation: 'localStorage',
            },
        });

        msal.handleRedirectCallback(error => {
            if (error) {
                console.error('Redirect callback error', error);
                return;
            }

            const request = { scopes: AZURE_DETAILS.scopes };

            msal.acquireTokenSilent(request)
                .then(accessTokenResponse => {
                    console.debug('Silent response');
                    handleAccessToken(accessTokenResponse.accessToken);
                })
                .catch(error => {
                    console.error('Failed to acquire token silently', error);
                    msal.acquireTokenRedirect(request);
                });
        });

        setMsalInstance(msal);
    }, []);

    const startLogin = useCallback(() => {
        getAccessToken(msalInstance);
    });

    useEffect(() => {
        const accessTokenParam = new URLSearchParams(
            window.location.search
        ).get('accessToken');

        if (accessTokenParam) {
            handleAccessToken(accessTokenParam);
        }
    }, []);

    return (
        <div className="Login">
            <div className="box" onClick={startLogin}>
                <img src={logo}></img>
                <h2>Sign in with SSO</h2>
            </div>
        </div>
    );
}
