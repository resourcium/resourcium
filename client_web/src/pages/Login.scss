$buttonBackground: #9B59B6;

.Login {
  min-width: 100%;
  min-height: 100%;
  background-color: #34495E;

  .box {
   position: absolute;
   top: 50%;
   left: 50%;
   transform: translate(-50%, -50%);
   display: flex;
   align-items: center;
   justify-content: center;

   flex-wrap: wrap;

   max-width: 600px;
   width: 95%;

   background-color: $buttonBackground;

   border-radius: 5px;
   overflow: hidden;

   cursor: pointer;

   transition: background-color 100ms linear;

   img {
     max-width: 150px;
     flex-grow: 0;
   }

   h2 {
     margin: 10px;
     color: #E8E8E8;
     font-size: 30px;
     flex-grow: 1;
     text-align: center;
     font-weight: 500;
   }

   &:hover {
     background-color: adjust-color($buttonBackground, $lightness: -5);
   }
  }
}
