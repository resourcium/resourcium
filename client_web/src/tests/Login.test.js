import { create } from 'react-test-renderer';
import React from 'react';

import Login from 'src/pages/Login';

import { wrapProvider } from 'src/tests/util';

test('app snapshot matches', () => {
    expect(create(wrapProvider(<Login />))).toMatchSnapshot();
});
