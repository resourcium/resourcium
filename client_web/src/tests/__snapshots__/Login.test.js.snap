// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app snapshot matches 1`] = `
<div
  className="Login"
>
  <div
    className="box"
    onClick={[Function]}
  >
    <img
      src="ucl-logo.jpg"
    />
    <h2>
      Sign in with SSO
    </h2>
  </div>
</div>
`;
